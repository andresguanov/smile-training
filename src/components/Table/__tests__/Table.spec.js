import { shallowMount } from "@vue/test-utils"
import smTable from "@components/Table/Table.vue"
import smMarkupTable from "@components/MarkupTable/MarkupTable.vue"
import smPagination from "@components/Pagination/Pagination.vue"
import smInnerLoading from "@components/Loading/InnerLoading.vue"
import Vue from "vue"

let wrapper
let columnConfig
const $id = _ => _
const rows = [
  { id: 1, name: "Nacho", age: 33, skills: "Vue, Php" },
  { id: 2, name: "Lian", age: 25, skills: "Vue, React" },
  { id: 3, name: "Seba", age: 60, skills: "Node" },
  { id: 4, name: "Martin", age: 54, skills: "Php" },
  { id: 5, name: "Carlos", age: 18, skills: "Python" },
  { id: 6, name: "Ariana", age: 23, skills: "Javascript" },
  { id: 7, name: "Juan", age: 46, skills: "Php" },
  { id: 8, name: "Pedro", age: 55, skills: "Python" },
  { id: 9, name: "Diana", age: 66, skills: "Javascript" },
  { id: 10, name: "Zeta", age: 21, skills: "Php" },
  { id: 11, name: "Guillermo", age: 19, skills: "Python" },
  { id: 12, name: "Edgardo", age: 57, skills: "Javascript" }
]
const filterConfig = {
  name: {
    type: "input",
    attrs: {
      placeholder: "Nombre"
    }
  },
  age: {
    type: "input",
    attrs: {
      nativeType: "number",
      placeholder: "Edad"
    }
  },
  skills: {
    type: "datepicker",
    attrs: {
      placeholder: "Aptitudes"
    }
  }
}

describe("smTable", () => {
  beforeEach(() => {
    columnConfig = [
      { name: "id", firstToUpper: true, order: false, headerClass: "TEST" },
      {
        name: "name",
        label: "nombre:",
        firstToUpper: true,
        bodyAlign: "right",
        filterable: false
      },
      { name: "age", label: "edad", format: s => s.toUpperCase(), filterable: false },
      {
        name: "skills",
        label: "Aptitudes",
        headerAlign: "center",
        bodyAlign: "right",
        width: "70",
        headerStyle: {
          background: "lightblue",
          "font-size": "18px",
          fontWeight: "bold"
        },
        bodyClass: "PRUEBA",
        bodyStyle: {
          "font-size": "15px",
          fontWeight: "thin"
        },
        filterable: false
      }
    ]
  })

  //#region Basic Render
  it("renders default table", async () => {
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.vm.localRows).toMatchObject(rows)
    expect(wrapper.findComponent(smMarkupTable).exists()).toBe(true)
    expect(wrapper.findComponent(smPagination).exists()).toBe(true)
    expect(wrapper.findComponent(smInnerLoading).exists()).toBe(true)
    expect(wrapper.find(".filters").exists()).toBe(false)
    expect(wrapper.findAll("thead th").wrappers.length).toBe(columnConfig.length)
    await Vue.nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll("tbody tr").wrappers.length).toBeLessThanOrEqual(wrapper.vm.perPage)
  })
  //#endregion

  //#region Props
  it("sets actions column's header", () => {
    const actionsColHeadText = "Actions"
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        actionsColHeadText
      },
      scopedSlots: {
        actionsCol: "<p>row actions</p>"
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.find("thead tr th:last-child").text()).toBe(actionsColHeadText)
  })

  it("sets filters", () => {
    columnConfig[1].filterable = true
    columnConfig[2].filterable = true
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        filterConfig
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.find(".filters").exists()).toBe(true)
  })

  it("sets perPage", () => {
    const itemsPerPage = 12
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        itemsPerPage
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.vm.perPage).toBe(itemsPerPage)
  })

  it("disables paginate", () => {
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        paginate: false
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.find("tfoot").exists()).toBe(false)
  })

  it("enable remote data", async () => {
    const fn = jest.fn()
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        remoteData: true,
        itemsPerPage: 13
      },
      listeners: {
        change: fn,
        refresh: fn
      },
      mocks: {
        $id
      }
    })

    await Vue.nextTick()
    expect(wrapper.vm.getRows()).toMatchObject(rows)
    wrapper.findComponent(smPagination).vm.$emit("refresh")
    expect(fn).toHaveBeenCalledTimes(1)
  })

  it("sets rows length", () => {
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        rowsLength: 30
      },
      mocks: {
        $id
      }
    })

    expect(wrapper.vm.totalItems).toBe(30)
  })
  //#endregion

  //#region Methods
  it("changes page", async () => {
    const change = jest.fn()
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig
      },
      listeners: {
        change
      },
      mocks: {
        $id
      }
    })
    wrapper.vm.onChangePage(1)
    expect(change).toHaveBeenCalledTimes(1)
    expect(change).toHaveBeenCalledWith(1)
    await wrapper.setProps({
      remoteData: true
    })
    wrapper.vm.onChangePage(1)
    expect(change).toHaveBeenCalledTimes(2)
  })
  //#endregion
  //#region Slots
  it("filter", () => {
    columnConfig[1].filterable = true
    columnConfig[2].filterable = true
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig,
        filterConfig
      },
      data() {
        return {
          showFilters: true
        }
      },
      slots: {
        "filter-name": "filter-name-content"
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.find(".filters").html()).toContain("filter-name-content")
  })

  it("default", () => {
    wrapper = shallowMount(smTable, {
      propsData: {
        rows,
        columnConfig
      },
      slots: {
        default: "<b id='default-slot'>default-slot-content</b>"
      },
      scopedSlots: {
        actionsCol: "<p>slot-actions</p>"
      },
      mocks: {
        $id
      }
    })
    expect(wrapper.find("#default-slot").text()).toBe("default-slot-content")
  })

  //#endregion
})
