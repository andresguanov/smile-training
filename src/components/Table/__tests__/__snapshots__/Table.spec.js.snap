// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smTable renders default table 1`] = `
<div style="position: relative; display: table;">
  <!---->
  <div class="table-min-height-setter">
    <sm-markup-table-stub separator="horizontal" separatorcolor="#ededed" style="width: 100%;">
      <thead>
        <!---->
        <tr class="headers">
          <!---->
          <th class="th TEST"><span style="display: inline-block;">
              Id
            </span>
            <!---->
          </th>
          <th class="th"><span style="display: inline-block;">
              Nombre:
            </span>
            <!---->
          </th>
          <th class="th"><span style="display: inline-block;">
              EDAD
            </span>
            <!---->
          </th>
          <th class="th center" style="background: lightblue; font-size: 18px; font-weight: bold;"><span style="display: inline-block;">
              Aptitudes
            </span>
            <!---->
          </th>
          <!---->
        </tr>
      </thead>
      <tbody>
        <!---->
        <tr>
          <!---->
          <td class="td">
            1
          </td>
          <td class="td right">
            Nacho
          </td>
          <td class="td">
            33
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Vue, Php
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            2
          </td>
          <td class="td right">
            Lian
          </td>
          <td class="td">
            25
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Vue, React
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            3
          </td>
          <td class="td right">
            Seba
          </td>
          <td class="td">
            60
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Node
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            4
          </td>
          <td class="td right">
            Martin
          </td>
          <td class="td">
            54
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Php
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            5
          </td>
          <td class="td right">
            Carlos
          </td>
          <td class="td">
            18
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Python
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            6
          </td>
          <td class="td right">
            Ariana
          </td>
          <td class="td">
            23
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Javascript
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            7
          </td>
          <td class="td right">
            Juan
          </td>
          <td class="td">
            46
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Php
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            8
          </td>
          <td class="td right">
            Pedro
          </td>
          <td class="td">
            55
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Python
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            9
          </td>
          <td class="td right">
            Diana
          </td>
          <td class="td">
            66
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Javascript
          </td>
          <!---->
        </tr>
        <tr>
          <!---->
          <td class="td">
            10
          </td>
          <td class="td right">
            Zeta
          </td>
          <td class="td">
            21
          </td>
          <td class="td right PRUEBA" style="font-size: 15px; font-weight: thin;">
            Php
          </td>
          <!---->
        </tr>
      </tbody>
    </sm-markup-table-stub>
  </div>
  <tfoot>
    <tr>
      <td colspan="4">
        <sm-pagination-stub value="0" itemsperpage="10" total="12" showrefresh="true" texts="[object Object]" style="margin: 0px auto;"></sm-pagination-stub>
      </td>
    </tr>
  </tfoot>
  <sm-inner-loading-stub text="Cargando..." style="position: absolute; top: 0px; left: 0px;"></sm-inner-loading-stub>
</div>
`;
